pkg/agent/plugin/nodeattestor/gcp/instanceIdentityToken.go

```diff
@@ -0,0 +1,178 @@
+	defaultIdentityTokenURL = "http://metadata/computeMetadata/v1/instance/service-accounts/default/identity"
+	queryParameters         = "?audience=%v&format=full"
+)
+
+type InstanceIdentityTokenAttestorConfig struct {
+	TrustDomain      string `hcl:"trust_domain"`
+	Audience         string `hcl:"audience"`
+	IdentityTokenURL string `hcl:"identity_troken_url"`
+}
+
+type InstanceIdentityTokenAttestorPlugin struct {
+	trustDomain      string
+	audience         string
+	identityTokenURL string
+
+	mtx *sync.RWMutex
+}
+
+func (p *InstanceIdentityTokenAttestorPlugin) spiffeID(gcpAccountID string, gcpInstanceID string) *url.URL {
+	spiffePath := path.Join("spire", "agent", pluginName, gcpAccountID, gcpInstanceID)
+	id := &url.URL{
+		Scheme: "spiffe",
+		Host:   p.trustDomain,
+		Path:   spiffePath,
+	}
+	return id
+}
+
+func httpGetBytes(url string) ([]byte, error) {
+	client := &http.Client{}
+	req, err := http.NewRequest("GET", url, nil)
+	if err != nil {
+		return nil, err
+	}
+	req.Header.Set("Metadata-Flavor", "Google")
+
+	resp, err := client.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+	bytes, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, err
+	}
+	return bytes, nil
+}
+
+func (p *InstanceIdentityTokenAttestorPlugin) FetchAttestationData(ctx context.Context, req *nodeattestor.FetchAttestationDataRequest) (*nodeattestor.FetchAttestationDataResponse, error) {
+	p.mtx.RLock()
+	defer p.mtx.RUnlock()
+
+	docBytes, err := httpGetBytes(fmt.Sprintf(p.identityTokenURL+queryParameters, p.audience))
+	if err != nil {
+		err = gcp.AttestationStepError("retrieving the Token from GCP", err)
+		return &nodeattestor.FetchAttestationDataResponse{}, err
+	}
+
+	parts := strings.Split(string(docBytes), ".")
+	iidAttestData := &cgcp.IIDAttestedData{}
+	header, err := base64.RawURLEncoding.DecodeString(parts[0])
+	if err != nil {
+		err = gcp.AttestationStepError(fmt.Sprintf("base64 decoding the IdentityToken header: %v", parts[0]), err)
+		return &nodeattestor.FetchAttestationDataResponse{}, err
+	}
+	iidAttestData.Header = string(header)
+	token, err := base64.RawURLEncoding.DecodeString(parts[1])
+	if err != nil {
+		err = gcp.AttestationStepError("base64 decoding the IdentityToken token", err)
+		return &nodeattestor.FetchAttestationDataResponse{}, err
+	}
+	iidAttestData.Token = string(token)
+	signature, err := base64.RawURLEncoding.DecodeString(parts[2])
+	if err != nil {
+		err = gcp.AttestationStepError("base64 decoding the IdentityToken signature", err)
+		return &nodeattestor.FetchAttestationDataResponse{}, err
+	}
+	iidAttestData.Signature = signature
+
+	var identityToken cgcp.IdentityToken
+	err = json.Unmarshal([]byte(iidAttestData.Token), &identityToken)
+	if err != nil {
+		err = gcp.AttestationStepError(fmt.Sprintf("unmarshalling the IdentityToken %v", string(docBytes)), err)
+		return &nodeattestor.FetchAttestationDataResponse{}, err
+	}
+
+	respData, err := json.Marshal(iidAttestData)
```

Comment:
The general flow here of "deconstruct, reconstruct" feels kind of roundabout, and worries me a little from a security perspective. Feels like it will be best to pass the token data to server side unmodified.

Maybe a shared function: func ParseGCPToken(string) (*Header, *IdentityToken, []byte)

Does it make sense to use an established JWT library for this? I noticed we end up having to reconstruct a part of the original token on server side for validation...

Reply:
The deconstruct, reconstruct approaches comes from the requirement that the payload needs the deconstructed and unmarshalled on the client in order to build the spiffeid and then needing to construct the JWT again to signature validation on the server.

Reply:
I'll look into established JWT libraries